src/coreclr/jit/lclmorph.cpp

```diff
@@ -826,6 +826,74 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
 
         switch (node->OperGet())
         {
+            case GT_INDEX_ADDR:
+            {
+                assert(TopValue(2).Node() == node);
+                assert(TopValue(1).Node() == node->gtGetOp1());
+                assert(TopValue(0).Node() == node->gtGetOp2());
+
+                if (node->gtGetOp1()->OperIs(GT_LCL_VAR, GT_LCL_ADDR) && node->gtGetOp2()->IsCnsIntOrI())
+                {
+                    if (TopValue(1).IsAddress() &&
+                        m_compiler->lvaGetDesc(TopValue(1).LclNum())->GetLayout()->IsBlockLayout())
+                    {
+                        ssize_t offset = node->AsIndexAddr()->gtElemOffset +
+                                         node->gtGetOp2()->AsIntCon()->IconValue() * node->AsIndexAddr()->gtElemSize;
+
+                        if (offset < static_cast<ssize_t>(m_compiler->lvaLclSize(TopValue(1).LclNum())))
+                        {
+                            if (FitsIn<unsigned>(offset) &&
+                                TopValue(2).AddOffset(TopValue(1), static_cast<unsigned>(offset)))
+                            {
+                                INDEBUG(TopValue(0).Consume());
+                                PopValue();
+                                PopValue();
+                                break;
+                            }
+                        }
+                        else
+                        {
+                            GenTree* gtThrow =
+                                m_compiler->gtNewMustThrowException(CORINFO_HELP_RNGCHKFAIL, node->TypeGet(),
+                                                                    node->AsIndexAddr()->gtStructElemClass);
+                            m_compiler->lvaGetDesc(gtThrow->AsOp()->gtOp2->AsLclVarCommon())->incLvRefCnt(1, RCS_EARLY);
+                            *use           = gtThrow;
+                            m_stmtModified = true;
+                        }
+                    }
+                }
+
+                EscapeValue(TopValue(0), node);
+                PopValue();
+                EscapeValue(TopValue(0), node);
+                PopValue();
+                break;
+            }
+            case GT_ARR_LENGTH:
+            {
+                assert(TopValue(1).Node() == node);
+                assert(TopValue(0).Node() == node->gtGetOp1());
+
+                if (node->gtGetOp1()->OperIs(GT_LCL_VAR, GT_LCL_ADDR))
+                {
+                    if (TopValue(0).IsAddress() &&
+                        m_compiler->lvaGetDesc(TopValue(0).LclNum())->GetLayout()->IsBlockLayout())
+                    {
+                        GenTree* gtLclFld =
+                            m_compiler->gtNewLclFldNode(TopValue(0).LclNum(), TYP_INT, OFFSETOF__CORINFO_Array__length);
```

Comment:
This should account in some way for `TopValue(0).Offset()` -- either assert that it is zero, or add it to the offset.